### PR TITLE
hotplug_unplug_during_io_repeat: New a case for testing plug

### PR DESCRIPTION
**Pull request by YongxueHong**
New a case that repeated hotplug/hot-unplug while
the device is busy serving I/O.

ID: 1231139
Signed-off-by: Yongxue Hong <yhong@redhat.com>

### CODE REVIEW

**Comment by YongxueHong** [comment]
Depends on https://github.com/autotest/tp-qemu/pull/1735

**Comment by peixiu** [comment]
@YongxueHong tested with patch 1735 together this one, failed as follows error:

23:04:22 ERROR| 
23:04:22 ERROR| Reproduced traceback from: /usr/local/lib/python3.6/site-packages/avocado_framework_plugins_vt-71.0-py3.6.egg/avocado_vt/test.py:438
23:04:22 ERROR| Traceback (most recent call last):
23:04:22 ERROR|   File "/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/hotplug_unplug_during_io_repeat.py", line 63, in run
23:04:22 ERROR|     iozone_thread = _run_iozone_background()
23:04:22 ERROR|   File "/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/hotplug_unplug_during_io_repeat.py", line 41, in _run_iozone_background
23:04:22 ERROR|     _check_iozone_status()
23:04:22 ERROR|   File "/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/hotplug_unplug_during_io_repeat.py", line 32, in _check_iozone_status
23:04:22 ERROR|     test.fail("Iozone is not alive!")
23:04:22 ERROR|   File "/usr/local/lib/python3.6/site-packages/avocado_framework-71.0-py3.6.egg/avocado/core/test.py", line 1046, in fail
23:04:22 ERROR|     raise exceptions.TestFail(message)
23:04:22 ERROR| avocado.core.exceptions.TestFail: Iozone is not alive!

File: qemu/tests/hotplug_unplug_during_io_repeat.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            logging.info('Start to run testing.(iteration: %d).' % (i + 1))
+            plug.hotplug_devs_serial()
+            if need_format:
+                mount_point = utils_disk.configure_empty_disk(
+                    session, plug[0], params['image_size_stg0'], os_type)[0]
+                if os_type == 'windows':
+                    need_format = False
+            iozone_thread = _run_iozone_background()
+            time.sleep(float(params['sleep_time']))
+            _check_iozone_status()
```
Comment (peixiu) [inline qemu/tests/hotplug_unplug_during_io_repeat.py]: for 63-65 lines, I saw the func _check_iozone_status() has called in _run_iozone_background(), it seems no need to check it again here.

File: qemu/tests/hotplug_unplug_during_io_repeat.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    :param test:   QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env:    Dictionary with test environment.
+    """
+    def _check_iozone_status():
+        ck_session = vm.wait_for_login(timeout=360)
+        if not utils_misc.wait_for(
+                lambda: check_cmd[os_type].split()[-1] in ck_session.cmd_output(
+                    check_cmd[os_type]), 180, step=3.0):
```
Comment (peixiu) [inline qemu/tests/hotplug_unplug_during_io_repeat.py]: it's wired that I tested found the fact waiting time is 1800 secs:
23:05:08 ERROR| ERROR 1-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.hotplug_unplug_during_io_repeat.q35 -> TestError: Not all IOZONE.EXE processes done in 1800 sec.

Comment (kimi1978wy) [reply qemu/tests/hotplug_unplug_during_io_repeat.py]: Hit the same error

Comment (YongxueHong) [reply qemu/tests/hotplug_unplug_during_io_repeat.py]: Hi, @kimi1978wy  @peixiu  I have updated the codes, pls help me to review it, thanks.
CC @XueqiangWei @zhenyzha

**Comment by zhenyzha** [comment]
(1/2) Host_RHEL.m8.u1.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.hotplug_unplug_during_io_repeat: ERROR: QMP command 'device_add' failed    (arguments: OrderedDict([('driver', 'virtio-blk-pci'), ('id', 'stg0'), ('drive', 'drive_stg0'), ('bus', 'pci.0')]),    error message: {'class': 'GenericError', 'desc': "Property 'virtio-blk-device.drive' can't find value... (323.58 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.hotplug_unplug_during_io_repeat: PASS (14315.79 s)

**Comment by YongxueHong** [comment]
Hi, @XueqiangWei  @peixiu  @kimi1978wy  There are some issues on q35, pls don't review, sorry.

**Comment by zhenyzha** [comment]
(1/2) Host_RHEL.m8.u1.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.hotplug_unplug_during_io_repeat: PASS (15647.84 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.hotplug_unplug_during_io_repeat: PASS (14071.72 s)
LGTM
Acked-by: Zhenyu Zhang <zhenyzha@redhat.com>

File: qemu/tests/cfg/hotplug_unplug_during_io_repeat.cfg [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    images += " stg0"
+    boot_drive_stg0 = no
+    image_name_stg0 = images/storage0
+    image_size_stg0 = 30G
+    remove_image_stg0 = yes
+    force_create_image_stg0 = yes
+    kill_vm_on_error = yes
+    iozone_timeout = 7200
+    sleep_time = 30
+    repeat_time = 300
```
Comment (peixiu) [inline qemu/tests/cfg/hotplug_unplug_during_io_repeat.cfg]: hi yhong, discussed with feature owners,  repeat_time = 100 is ok.

Comment (YongxueHong) [reply qemu/tests/cfg/hotplug_unplug_during_io_repeat.cfg]: ok , have updated it, thanks.

File: qemu/tests/cfg/hotplug_unplug_during_io_repeat.cfg [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    vt_ulimit_nofile = 10240
+    q35:
+        pcie_extra_root_port = 1
+    Linux:
+        iozone_options = '-az -g 10G -y 32k -i 0 -i 1 -I -f {0}/iozone_test'
+    Windows:
+        iozone_options = '-az -b C:\{0}_stress_test -g 10G -y 32k -i 0 -i 1 -I -f {0}:\iozone_test'
+    variants:
+        - @default:
+        - with_scsi_hba:
```
Comment (kimi1978wy) [inline qemu/tests/cfg/hotplug_unplug_during_io_repeat.cfg]: For virtio-blk-pci, windows  cannot support unplug controller when running iozone.

Comment (YongxueHong) [reply qemu/tests/cfg/hotplug_unplug_during_io_repeat.cfg]: > For virtio-blk-pci, windows cannot support unplug controller when running iozone.
Yes, so it is only for virtio_scsi here.

Comment (YongxueHong) [reply qemu/tests/cfg/hotplug_unplug_during_io_repeat.cfg]: Thanks, will update it.

Comment (XueqiangWei) [reply qemu/tests/cfg/hotplug_unplug_during_io_repeat.cfg]: unplug device in use scsi-cd/hd/xxx is valid, virtio-scsi-pci is not valid for windows guest

**Comment by kimi1978wy** [comment]
LGTM and test pass
Acked by: Yu Wang <wyu@redhat.com>

 (1/1) Host_RHEL.m8.u1.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.hotplug_unplug_during_io_repeat.q35: PASS (5438.66 s) 
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.hotplug_unplug_during_io_repeat.q35: PASS (4021.63 s）

**Comment by peixiu** [comment]
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2008.x86_64.r2.io-github-autotest-qemu.hotplug_unplug_during_io_repeat.q35: PASS (5100.98 s)

Acked-by: Peixiu Hou phou@redhat.com

**Comment by XueqiangWei** [comment]
'1-Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.hotplug_unplug_during_io_repeat' Finshed 2019-10-16 08:33:26 2019-10-16 09:55:06 PASS 4900.061079
'2-Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.hotplug_unplug_during_io_repeat.with_scsi_hba' Finshed 2019-10-16 09:55:06 2019-10-16 11:17:18 PASS 4931.959317
'3-Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.hotplug_unplug_during_io_repeat' Finshed 2019-10-16 11:17:18 2019-10-16 12:28:37 PASS 4278.536112

LGTM,
Acked-by xuwei@redhat.com

**Comment by YongxueHong** [comment]
@lolyu  help to review, thanks.

**Comment by YongxueHong** [comment]
Thanks all, Let's merge it.

**Pull request merged**